### PR TITLE
New version of pattern open, separating local equations in Env.t

### PR DESCRIPTION
**Pull request by garrigue**
This is a new branch of GPR #187, which reduces the amount of changes by putting local constrains in a specific field of Env.t rather than mixing them with types.
This should help for other extensions.

Note that a large part of the changes are due to commit 68fdc8a, adding a compare function to the Path module, which should have been done ages ago :)

### CODE REVIEW

File: bytecomp/lambda.ml [hunk 1]
```diff
@@ -431,11 +431,7 @@ let iter f = function
       f e
 
 
-module IdentSet =
-  Set.Make(struct
-    type t = Ident.t
-    let compare = compare
-  end)
+module IdentSet = Set.Make(Ident)
```
Comment (Drup) [inline bytecomp/lambda.ml:434]: `Ident.Set` ? (added with all the `Identifiable` stuff with flambda)

**Comment by alainfrisch** [comment]
Currently, the constraints are only on simple identifiers, not qualified paths, right?  I guess you directly introduced a PathMap to support gadt constraints on non-local abstract types (for http://caml.inria.fr/mantis/view.php?id=7233 ).  Is that right?

File: bytecomp/lambda.ml (hunk 1 above)
Comment (garrigue) [inline bytecomp/lambda.ml:434]: Fine, but this is completely orthogonal (the code was already there).

**Comment by garrigue** [comment]
Yes, and this is also needed for modular implicits.
But I'm not sure it is that easy... currently this only works for find_type, not lookup_type.
A solution would be to always use find_type after lookup_type, like for lookup_module / find_module.
Will think about that later.

**Comment by gasche** [comment]
I merged in trunk ( 2d33e164a6b636caa5207b765f427360ffb14778 ), thanks!
